In [8]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import scale
from sklearn.model_selection import KFold
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import cross_val_score

#вместо того что указано в ПДФ, так как load_boston() удален из sklearn
data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep=r"\s+", skiprows=22, header=None)
data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]

2. Приведите признаки в выборке к одному масштабу при помощи
функции sklearn.preprocessing.scale.

In [11]:
data_scaled = scale(data)

3. Переберите разные варианты параметра метрики p по сетке от 1 до
10 с таким шагом, чтобы всего было протестировано 200 вариантов
(используйте функцию numpy.linspace).Используйте KNeighborsRegressor
сn_neighbors = 5 и weights = ’distance’—данный параметр добавляет
 в алгоритм веса,зависящие от расстояния до ближайших соседей.В
качестве метрики качества используйте среднеквадратичную ошиб-
ку (параметр scoring=’mean_squared_error’ у cross_val_score; при
использовании библиотеки scikit-learn версии 18.0.1 и выше необхо-
димоуказыватьscoring=’neg_mean_squared_error’).Качество оце-
нивайте, как и в предыдущем задании, с помощью кросс-валидации
по 5 блокам с random_state = 42, не забудьте включить перемеши-
вание выборки (shuffle=True).

In [15]:
cv = KFold(n_splits=5, shuffle=True, random_state=42)
best_mean = -np.inf
best_p = 0
for p in np.linspace(1, 10, num=200):
    model = KNeighborsRegressor(n_neighbors=5, weights = 'distance', p = p)
    scores = cross_val_score(model, data, target, cv = cv, scoring='neg_mean_squared_error')
    mean = scores.mean()
    if mean > best_mean:
        best_mean = mean
        best_p = p
print(best_p)
print(-best_mean)

1.0
29.7872280754304


4. Определите, при каком p качество на кросс-валидации оказалось
оптимальным. Обратите внимание, что cross_val_score возвраща-ет массив показателей качества по блокам; необходимо максимизи-
ровать среднее этих показателей. Это значение параметра и будет
ответом на задачу.

Оптимальное качество кросс-валидации при p = 1, а минимальная среднеквадратичная оценка = 29.8